[PriyadarshiCharvi] iP

PriyadarshiCharvi opened the pull request:
### ORACLE ### 

> “The best way to predict the future is to create it.” – Peter Drucker

_**Oracle**_ frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is:

1. Download it from [here](https://nus-cs2103-ay2425s2.github.io/website/schedule/week4/project.html#:~:text=download%20it%20from-,here)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 

And it is **FREE**! 🎉

Features:

- [X] Managing Tasks
- [X] Managing Deadlines
- [ ] Managing Events
- [ ] An attractive user interface (Coming soon! 🔜 )

If you want to try it out, you can follow the following steps:

1. Download the jar file in a directory of your choice, say, 'oracle-jar'
2. Run the following commands:

```
# Step 1: Navigate to the directory containing the file
cd oracle-jar

#Step 2: Run the jar file
java -jar Oracle.jar

```

_**Enjoy!!**_

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename chatbot

Commit:
Add echo functionality

Commit:
Implement task storage and list functionality

Commit:
Add mark/unmark feature

Commit:
Add todo, event and deadline tasks

Commit:
Add automated testing

Commit:
Add errors for invalid arguments

Commit:
Add delete functionality for tasks

Commit:
Add enums

Commit:
Implement task storage functionality

Commit:
Merge branch 'branch-Level-7'

Commit:
Add date and time support

Commit:
Merge branch 'branch-Level-8'

Rae2480 commented on src/main/java/Oracle.java:
```diff
@@ -0,0 +1,381 @@
+import java.io.IOException;
```
Hi! Generally, and also according to 2103 instructions, all classes should belong to a package. That seems to be missing here! Is this intentional?

Rae2480 commented on src/main/java/Oracle.java:
```diff
@@ -0,0 +1,381 @@
+            case EVENT -> "[E]";
+        };
+        return typeIcon + "[" + getStatusIcon() + "] " + description;
+    }
```
I like the way you've used enums! Its very clean and also easy to understand! ☺️

Rae2480 commented on src/main/java/Oracle.java:
```diff
@@ -0,0 +1,381 @@
+    }
+}
+
+class Deadline extends Task {
```
Do you think these different classes could be in separate files? It might make the code on each page less lengthy!

Rae2480 commented on src/main/java/Oracle.java:
```diff
@@ -0,0 +1,381 @@
+                            break;
+                        default:
+                            continue;
+                    }
```
switch & case statements should be aligned. most IDEs automatically indent case, but you can change this configuration to make it easier on yourself & not have to manually fix this every time!

the exception to the rule is with lambda-style switch-case statements fyi!

Rae2480 commented on src/main/java/Oracle.java:
```diff
@@ -0,0 +1,381 @@
+        }
+    }
+
+    public Deadline(String description, LocalDateTime by) {
```
do you think you could add header comments for all the public classes and methods, such as this? 

it would help readers understand what ur code does, even when they dont understand the code exactly

Rae2480 commented on src/main/java/Oracle.java:
```diff
@@ -0,0 +1,381 @@
+            for (String line : Files.readAllLines(DATA_FILE)) {
+                try {
+                    String[] parts = line.split("\\|");
+                    if (parts.length < 3) continue;
```
not entirely sure if this is "wrong" per se. but it might look more consistent across the board if this continue statement was in the traditional if-statement format (with the curly brackets and indentation)

Rae2480 commented on src/main/java/Oracle.java:
```diff
@@ -0,0 +1,381 @@
+                        oracle.tasks.get(index).markDone();
+                        oracle.saveTasks();
+                        System.out.println("    ____________________________________________________________");
+                        System.out.println("    Great! I've marked this task as done:");
+                        System.out.println("    " + oracle.tasks.get(index));
+                        System.out.println("    ____________________________________________________________");
+                } else if (input.startsWith("unmark")) {
+                    int index = Integer.parseInt(input.substring(7)) - 1;
+                    if (index < 0 || index > oracle.tasks.size()) {
+                        throw new OracleException("OOPS!!! Task number is invalid. Please enter a number between 1 and " + oracle.tasks.size());
+                    }
+                        oracle.tasks.get(index).markUndone();
+                        oracle.saveTasks();
+                        System.out.println("    ____________________________________________________________");
+                        System.out.println("    Alright, I've marked this task as not done yet:");
+                        System.out.println("    " + oracle.tasks.get(index));
+                        System.out.println("    ____________________________________________________________");
+                } else if (input.equals("bye")) {
+                    System.out.println("    ____________________________________________________________");
+                    System.out.println("    Bye. Hope to see you again soon!");
+                    System.out.println("    ____________________________________________________________");
+                    break;
+                } else {
+                    throw new OracleException("OOPS!!! I'm sorry, but I don't know what that means :-(. Try something like 'todo assignment'.");
+                }
+            } catch (OracleException e) {
+                System.out.println("    ____________________________________________________________");
+                System.out.println("    " + e.getMessage());
+                System.out.println("    ____________________________________________________________");
+            } catch (NumberFormatException e) {
+                System.out.println("    ____________________________________________________________");
+                System.out.println("    Please enter a valid task number.");
+                System.out.println("    ____________________________________________________________");
+            } catch (Exception e) {
+                System.out.println("    ____________________________________________________________");
+                System.out.println("    OOPS!!! " + e.getMessage());
+                System.out.println("    ____________________________________________________________");
+            }
+
+        }
```
could you extract out most of this into classes? each input could be handled separately then (separation of concerns and all)

Rae2480 commented on src/main/java/Oracle.java:
```diff
@@ -0,0 +1,381 @@
+                    }
+                    int index = Integer.parseInt(input.substring(7)) - 1;
+                    if (index < 0 || index >= oracle.tasks.size()) {
+                        throw new OracleException("Invalid task number. Please enter a number between 1 and " + oracle.tasks.size());
```
i think this line (and a few others in this method body) might be too lengthy as it goes down to the line below. you could consider a line break? don't forget to index 4 extra spaces 😉

Rae2480 reviewed (commented):
Overall, the logic of your code makes sense so good job! I personally feel that it's a bit crowded all in this one file. separating out into different files and then further into different packages might help make your code more readable! Otherwise, good job so far and all the best with future weeks of iP increments 👍

Commit:
Add more OOP

Commit:
Merge branch 'branch-A-MoreOOP'

Commit:
Add packages

Commit:
Merge tag 'A-Packages'

Commit:
Merge branch 'add-gradle-support'

curiousfun88 commented on src/main/java/oracle/common/Storage.java:
```diff
@@ -0,0 +1,114 @@
+package oracle.common;
+
+import oracle.task.*;
```
Would it be better to name the specific class(es) that you want to import? (Avoid wildcards import)

curiousfun88 commented on src/main/java/oracle/common/Storage.java:
```diff
@@ -0,0 +1,114 @@
+                    Task task;
+
+                    switch (type) {
+                        case "T":
```
Would it be better to not indent case (ie. switch and case have same indentation) to follow the coding standard required?

curiousfun88 commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,45 @@
-|____/ \__,_|_|\_\___|
-
+    ____________________________________________________________
+    Hello! I'm oracle.Oracle
```
Would it be better to replace oracle.Oracle with just Oracle, to make it sound more natural? :)

curiousfun88 commented on src/main/java/oracle/task/Task.java:
```diff
@@ -0,0 +1,42 @@
+
+    public Task(String description, TaskType type) {
+        if (description.isBlank()) {
+            throw new IllegalArgumentException("oracle.task.Task description cannot be empty.");
```
Would it be clearer and easier to understand for the exception message to be just "Task description..." instead of "oracle.task.Task..."?

curiousfun88 reviewed (commented):
The code is well-organized and easy to understand. It was a pleasure reviewing your code! Good job and all the best! LGTM!!

crmlatte commented on src/main/java/oracle/command/UnmarkCommand.java:
Command classes look ok!

crmlatte commented on src/main/java/oracle/common/Parser.java:
```diff
@@ -0,0 +1,124 @@
+        } else {
+            throw new OracleException("OOPS!!! I'm sorry, but I don't know what that means :-(. Try something like 'todo assignment'.");
+        }
+    }
```
Maybe can use a switch statement? instead of so many if else

crmlatte commented on src/main/java/oracle/common/Storage.java:
```diff
@@ -0,0 +1,114 @@
+                    String description = parts[2].trim();
+                    Task task;
+
+                    switch (type) {
```
The indentation for switch and case should be the same!

crmlatte commented on src/main/java/oracle/task/Event.java:
Nice, variable and method names look clear!

crmlatte reviewed (commented):
Overall ur code looks nice, maybe just a few coding standard needs to be fixed! Also remember to add java docs in all ur classes!

Commit:
Fix coding standards

Commit:
Add JUnit tests

Commit:
Update gradle configurations

Commit:
Revert "Update gradle configurations"

This reverts commit 698c72d8aa0ee5876730e75f0301f4d1ece3ec19.

Commit:
Update configuration for jar support

Commit:
Add JavaDocs

Commit:
Apply coding standards

Commit:
Implement find feature

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

elizazaa commented on src/main/java/oracle/command/DeleteCommand.java:
```diff
@@ -0,0 +1,40 @@
+     * @throws OracleException If an error occurs while deleting the task or saving to storage.
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws OracleException {
```
Good SLAP

elizazaa reviewed (commented):
LGTM!

Commit:
Add checkstyle configuration

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Fix checkstyle errors

Commit:
Implement basic GUI using JavaFX

Commit:
Fix javadoc

Commit:
Merge branch 'branch-Level-10'

Commit:
Refactor methods to use varargs

Commit:
Add assertion checks for important preconditions

Assertions help catch logical errors early by ensuring that important
assumptions hold at runtime. The current implementation lacks assertion
checks, which may result in undetected errors leading to unexpected behaviour.

To improve code robustness, let's:
- Add assertions in `TaskList.java` to ensure valid task indices before accessing or deleting tasks.
- Add assertions in `Parser.java` to check that input commands are not null or empty.
- Ensure valid file paths and task formats in `Storage.java` before reading/writing.

These changes will help detect potential issues earlier and improve overall
code reliability.

PriyadarshiCharvi committed:
Merge pull request #2 from PriyadarshiCharvi/branch-A-Assertions

Add assertion checks for important preconditions

Commit:
Refactor code for better readability and maintainability

Long methods, deep nesting, and redundant logic made parts of the code difficult to read and modify.

Let's:

- Refactor parseDeadlineCommand() for clearer validation and string handling.
- Reorganize load() in Storage by extracting helper methods to reduce nesting.
- Replace if-else conditions in parse() with switch-case for better readability.

These improvements enhance code maintainability while preserving functionality, ensuring a cleaner and more efficient codebase.

Commit:
Use java streams for some methods

Commit:
Merge branch 'branch-A-CodeQuality'

Commit:
Merge branch 'branch-A-Streams'

Commit:
Refactor code to use Java Streams for improved efficiency

Several parts of the code relied on traditional loops, which led to verbose and less readable logic.

So to fix that, let's

- Refactor findTasks() in TaskList to use Stream.filter() for better readability and efficiency.
- Simplify parseTasks() in Storage by utilizing stream().map().collect() to process task entries.
- Replace manual iteration in showTasks() with forEach() for cleaner output handling.

Using Java Streams enhances code conciseness and optimises performance while maintaining the intended functionality.

PriyadarshiCharvi committed:
Merge pull request #3 from PriyadarshiCharvi/branch-A-Streams

Refactor code to use Java Streams for improved efficiency

Commit:
Implement Continuous Integration with GitHub Actions

Commit:
Implement the snooze functionality to reschedule events or deadlines

Commit:
Improve GUI responsiveness and dynamic resizing for better UX

Commit:
Merge branch 'branch-BetterGUI'

Commit:
Give a space-themed personality to Oracle

Commit:
Merge branch 'branch-Personality'

Commit:
Implement the help command for users to see list of commands

Commit:
Merge branch 'branch-HelpCommand'

Commit:
Improve GUI colours and resizing

PriyadarshiCharvi committed:
Update README.md

Commit:
Update user guide and window screenshot

PriyadarshiCharvi committed:
Update README.md
